In [1]:
from shapely.geometry import Point
import pandas as pd
import pickle
import matplotlib.pyplot as plt
from datetime import date, timedelta, datetime
import geopandas as gpd

loading may data - pickle and csv


In [2]:
%%time
june_loc = pd.read_pickle("../data/june.pkl")
june_trip = pd.read_csv("../data/june_trip.csv")

Wall time: 5.32 s


In [3]:
june_trip['pubTimeStamp'] = pd.to_datetime(june_trip['pubTimeStamp'])


This section makes a dictionary of dataframes. The dataframes can be "chunked up" into different time segments. The main things you need to adjust: the start and end variables, and the delta variable, which designates how "big" a "time chunk" should be.

In [4]:
#make a list of datetimes with datetime function 
start = datetime(2019, 6, 8, 0, 0)
end = datetime(2019, 6, 9, 0, 0)
delta = timedelta(days=1)
timelist = []
timenamelist = []
while start <= end:
    timelist.append(start)
    timenamelist.append(start.strftime("%Y-%m-%d-%H-%M"))
    start += delta
# print(timelist)
# print(timenamelist)

In [5]:
#goal of function: using list of time frames, in this case, days, make a dataframe containing
#info about the location for each scooter, for each day


#make dictionary of dataframes using base dataframe and a user-defined segments_of_time_list
#from cell above
def make_day_df(datetime_list, timenames_list, dataframe, colname):
    resultslist = {} #turn into dictionary
    x = 0
    y = 1
    z = 0
    while x in range(0,len(datetime_list)-1):
        starttime = datetime_list[x]
        endtime = datetime_list[y]
        mask = (dataframe[colname] > starttime) & (dataframe[colname] <= endtime)
        framename = dataframe.loc[mask]
        x += 1
        y += 1
        print(x)
        resultslist[timenames_list[z]] = framename
        z += 1
    return resultslist
        
the_output = make_day_df(timelist, timenamelist, june_trip,"pubTimeStamp")


1


In [6]:
the_output.keys()

dict_keys(['2019-06-08-00-00'])

I'm interested in May 26th specifically, the Saturday of Memorial Day weekend, since it had the most scooter rides.

In [7]:
the_output['2019-06-08-00-00']
# the_output['2019-05-01-00-00']['sumdid'].nunique()

pubTimeStamp    companyName tripRecordNum  \
53908 2019-06-08 00:02:45.000  Bolt Mobility      BOL00208   
53909 2019-06-08 00:02:45.000  Bolt Mobility      BOL00209   
53910 2019-06-08 00:02:48.583           Lyft          LFT2   
53911 2019-06-08 00:03:05.680           SPIN        SPI491   
53912 2019-06-08 00:03:05.680           SPIN        SPI509   
...                       ...            ...           ...   
67547 2019-06-08 23:59:33.227           Bird       BRD8454   
67548 2019-06-08 23:59:33.227           Bird       BRD8466   
67549 2019-06-08 23:59:33.227           Bird       BRD8459   
67550 2019-06-08 23:59:33.227           Bird       BRD8452   
67551 2019-06-08 23:59:33.227           Bird       BRD8462   

                                             sumdID  tripDuration  \
53908  Powered-4820bbcf-520e-cfe9-b486-b24182a6874e    453.000000   
53909  Powered-d8df5c0c-0452-dd77-df85-53e92efa5d25   1612.000000   
53910                                 Powered223899      9.818517   
53911                                Powered5183517     49.000000   
53912                                Powered7341390     10.000000   
...                                             ...           ...   
67547                                  PoweredBMSNR      3.000000   
67548                                  PoweredAGR3R     12.000000   
67549                                  PoweredWXHD3      7.000000   
67550                                PoweredUNKNOWN      0.000000   
67551                                  PoweredHM4FI      8.000000   

       tripDistance            startDate        startTime  \
53908    2099.74000  2019-06-07 00:00:00         23:50:15   
53909   12217.85000  2019-06-07 00:00:00         23:31:27   
53910    6916.01072  2019-06-07 00:00:00  23:52:59.366666   
53911    4455.38072  2019-06-07 00:00:00         23:12:00   
53912     410.10500  2019-06-07 00:00:00         23:50:52   
...             ...                  ...              ...   
67547       0.00000  2019-06-08 00:00:00  23:57:19.010000   
67548    6233.59600  2019-06-08 00:00:00  23:49:20.373333   
67549    1312.33600  2019-06-08 00:00:00  23:53:40.570000   
67550       0.00000  2019-06-08 00:00:00  23:59:40.403333   
67551    2952.75600  2019-06-08 00:00:00  23:51:32.826666   

                   endDate          endTime  startLatitude  startLongitude  \
53908  2019-06-07 00:00:00         23:57:48      36.180128      -86.793786   
53909  2019-06-07 00:00:00         23:58:19      36.149888      -86.806880   
53910  2019-06-08 00:00:00  00:02:48.476666      36.148540      -86.806970   
53911  2019-06-08 00:00:00         00:00:16      36.151914      -86.781197   
53912  2019-06-08 00:00:00         00:00:18      36.164096      -86.779547   
...                    ...              ...            ...             ...   
67547  2019-06-09 00:00:00  00:00:03.513333      36.168200      -86.783100   
67548  2019-06-09 00:00:00  00:01:21.513333      36.175100      -86.782000   
67549  2019-06-09 00:00:00  00:00:46.890000      36.177100      -86.787800   
67550  2019-06-09 00:00:00  00:00:04.966666      36.126400      -86.789300   
67551  2019-06-09 00:00:00  00:00:02.826666      36.173500      -86.785500   

       endLatitude  endLongitude  \
53908    36.180059    -86.793833   
53909    36.158263    -86.777445   
53910    36.146750    -86.815700   
53911    36.164000    -86.779081   
53912    36.164977    -86.780421   
...            ...           ...   
67547    36.168000    -86.783000   
67548    36.167300    -86.777100   
67549    36.176900    -86.787600   
67550    36.126400    -86.789300   
67551    36.169300    -86.787200   

                                               tripRoute  \
53908  [["36.180128","-86.793786"],["36.180115","-86....   
53909  [["36.149888","-86.806880"],["36.149900","-86....   
53910  [[36.14854,-86.80697],[36.14854,-86.80703],[36...   
53911  [[36.16775,-86.78647],[36.16784,-86.78655],[36...   
53912  [[36.164976799145848,-86.78042142

In [8]:
the_output['2019-06-08-00-00']['sumdID'].value_counts()
# the_output['2019-05-01-00-00']['sumdid'].nunique()

PoweredUNKNOWN          158
PoweredIUDHOZW2ZZ725     18
PoweredSINZNV4OYDROU     16
Powered211524            16
PoweredC62USEYDMW5FF     16
                       ... 
Powered3V7DM              1
Powered84ZJC              1
PoweredEXGA9              1
PoweredADYBC              1
PoweredAGR3R              1
Name: sumdID, Length: 3707, dtype: int64

What I'd like to do is return only the Scooter IDs that had less than 3 rides on this day. Then be able to plot those points to see where they are clustered, because that would tell us where we can reduce the scooter density.

In [9]:
the_output['2019-06-08-00-00']['sumdID'].value_counts().loc[lambda x : x<3]

Powered2I3MS    2
PoweredZG428    2
PoweredR5S4G    2
PoweredJ2MAY    2
Powered3B46L    2
               ..
Powered3V7DM    1
Powered84ZJC    1
PoweredEXGA9    1
PoweredADYBC    1
PoweredAGR3R    1
Name: sumdID, Length: 1569, dtype: int64

So now we can see that on 6/8/19, 3549 (subtracting the row for unknowns) unique scooter IDs had rides. Of those 3878 scooters, 1569 only had 1 or 2 rides (not meeting the 3 rides per day threshold). So only 56% of scooters on this day had 3 or more rides.

I'm also curious to know what percentage of scooters are over-performing though. For instance, how many scooters are being used for at least 6 rides per day (twice the stated goal)?

In [10]:
the_output['2019-06-08-00-00']['sumdID'].value_counts().loc[lambda x : x>5]

PoweredUNKNOWN          158
PoweredIUDHOZW2ZZ725     18
PoweredSINZNV4OYDROU     16
Powered211524            16
PoweredC62USEYDMW5FF     16
                       ... 
Powered054202             6
PoweredCP47V              6
PoweredE6BZVWL4K3HI6      6
Powered675277             6
Powered879199             6
Name: sumdID, Length: 731, dtype: int64

So 573 unique scooters had twice (or more) the stated rides-per-day goal on this particular day. That's 16% of the scooters.

In [17]:
#make a list of datetimes with datetime function 
start = datetime(2019, 6, 28, 0, 0)
end = datetime(2019, 6, 29, 0, 0)
delta = timedelta(days=1)
timelist = []
timenamelist = []
while start <= end:
    timelist.append(start)
    timenamelist.append(start.strftime("%Y-%m-%d-%H-%M"))
    start += delta
# print(timelist)
# print(timenamelist)

In [18]:
#goal of function: using list of time frames, in this case, days, make a dataframe containing
#info about the location for each scooter, for each day


#make dictionary of dataframes using base dataframe and a user-defined segments_of_time_list
#from cell above
def make_day_df(datetime_list, timenames_list, dataframe, colname):
    resultslist = {} #turn into dictionary
    x = 0
    y = 1
    z = 0
    while x in range(0,len(datetime_list)-1):
        starttime = datetime_list[x]
        endtime = datetime_list[y]
        mask = (dataframe[colname] > starttime) & (dataframe[colname] <= endtime)
        framename = dataframe.loc[mask]
        x += 1
        y += 1
        print(x)
        resultslist[timenames_list[z]] = framename
        z += 1
    return resultslist
        
the_output = make_day_df(timelist, timenamelist, june_trip,"pubTimeStamp")


1


In [19]:
the_output.keys()

dict_keys(['2019-06-28-00-00'])

In [20]:
the_output['2019-06-28-00-00']
# the_output['2019-05-01-00-00']['sumdid'].nunique()

pubTimeStamp companyName tripRecordNum          sumdID  \
183970 2019-06-28 00:04:35.743        Bird       BRD2574    Powered6Y1U7   
183971 2019-06-28 00:05:28.530        SPIN        SPI256  Powered2763775   
183972 2019-06-28 00:05:28.530        SPIN          SPI3  Powered1199762   
183973 2019-06-28 00:06:03.367        Lyft          LFT1   Powered405856   
183974 2019-06-28 00:09:35.860        Bird       BRD2585    Powered8QBKJ   
...                        ...         ...           ...             ...   
189855 2019-06-28 23:57:52.477        Lyft        LFT901   Powered695890   
189856 2019-06-28 23:58:09.620        Lyft        LFT902   Powered787647   
189857 2019-06-28 23:58:37.993        Lyft        LFT903   Powered512435   
189858 2019-06-28 23:58:41.770        Lyft        LFT904   Powered130485   
189859 2019-06-28 23:58:47.287        Lyft        LFT905   Powered151217   

        tripDuration  tripDistance            startDate        startTime  \
183970      8.000000       0.00000  2019-06-27 00:00:00  23:57:42.680000   
183971      7.000000    3802.49356  2019-06-27 00:00:00         23:54:49   
183972      3.000000    1217.19164  2019-06-28 00:00:00         00:01:24   
183973     10.556183    3799.21272  2019-06-27 00:00:00  23:55:29.890000   
183974      5.000000       0.00000  2019-06-28 00:00:00  00:05:56.600000   
...              ...           ...                  ...              ...   
189855     10.609150    4484.90828  2019-06-28 00:00:00  23:47:15.843333   
189856      2.340683       0.00000  2019-06-28 00:00:00  23:55:49.060000   
189857      3.551983    3904.19960  2019-06-28 00:00:00  23:55:04.746666   
189858      8.068033    4222.44108  2019-06-28 00:00:00  23:50:37.576666   
189859      8.223883    4222.44108  2019-06-28 00:00:00  23:50:33.676666   

                    endDate          endTime  startLatitude  startLongitude  \
183970  2019-06-28 00:00:00  00:05:12.830000      36.165100      -86.781700   
183971  2019-06-28 00:00:00         00:01:23      36.140843      -86.807557   
183972  2019-06-28 00:00:00         00:03:35      36.148054      -86.807144   
183973  2019-06-28 00:00:00  00:06:03.263333      36.134410      -86.799930   
183974  2019-06-28 00:00:00  00:10:52.120000      36.164000      -86.779900   
...                     ...              ...            ...             ...   
189855  2019-06-28 00:00:00  23:57:52.393333      36.153450      -86.788990   
189856  2019-06-28 00:00:00  23:58:09.503333      36.164300      -86.769330   
189857  2019-06-28 00:00:00  23:58:37.863333      36.161320      -86.777270   
189858  2019-06-28 00:00:00  23:58:41.660000      36.167990      -86.769030   
189859  2019-06-28 00:00:00  23:58:47.106666      36.167950      -86.769030   

        endLatitude  endLongitude  \
183970    36.161200    -86.777500   
183971    36.148788    -86.799222   
183972    36.150757    -86.804722   
183973    36.131020    -86.800270   
183974    36.169000    -86.772100   
...             ...           ...   
189855    36.158980    -86.775900   
189856    36.164320    -86.769340   
189857    36.165700    -86.768380   
189858    36.160210    -86.776480   
189859    36.160220    -86.776480   

                                                tripRoute  \
183970  [[36.165083,-86.781663],[36.165183,-86.781641]...   
183971  [[36.1487881,-86.7992222],[36.14961,-86.80027]...   
183972  [[36.150756775023964,-86.804722209221467],[36....   
183973  [[36.13441,-86.79993],[36.13448,-86.7999],[36....   
183974  [[36.164107,-86.77976],[36.164198,-86.779757],...   
...                                                   ...   
189855  [[36.15345,-86.78899],[36.1535,-86.78896],[36....   
189856         [[36.1643,-86.76933],[36.16432,-86.76934]]   
189857  [[36.16132,-86.77727],[36.16124,-86.77714],[36...   
189858  [[36.16799,-86.76903],[36.16793,-86.76911],[36...   
189859  [[36.16795,-86.76903],[36.16788,-86.76909],[36...   

                         create_dt  
183970  2019-06-29 05:30:27.

In [21]:
the_output['2019-06-28-00-00']['sumdID'].value_counts()
# the_output['2019-05-01-00-00']['sumdid'].nunique()

PoweredUNKNOWN                                  42
Powered755BVAK3ZNJYW                            23
PoweredXZH7USFBU76DX                            21
PoweredOKXVAKQFFLXU5                            20
PoweredZ5IMPCBKO5P7E                            19
                                                ..
Powered6DJHC                                     1
Powered-a31dc125-5151-18b2-ac70-0301656484bd     1
Powered667863                                    1
PoweredDUDFB                                     1
Powered512435                                    1
Name: sumdID, Length: 2134, dtype: int64

In [22]:
the_output['2019-06-28-00-00']['sumdID'].value_counts().loc[lambda x : x<3]

Powered1N67J                                    2
PoweredXN2NR                                    2
Powered799T4                                    2
PoweredLXV7RJSLHE56P                            2
PoweredMDAQYMTPMLQL2                            2
                                               ..
Powered6DJHC                                    1
Powered-a31dc125-5151-18b2-ac70-0301656484bd    1
Powered667863                                   1
PoweredDUDFB                                    1
Powered512435                                   1
Name: sumdID, Length: 1411, dtype: int64

So out of 2092 unique scooters that ran on 6/28/19, 1411 of them had fewer than 3 rides. So only 33% of the scooters met the threshold of 3 rides that day.

In [23]:
the_output['2019-06-28-00-00']['sumdID'].value_counts().loc[lambda x : x>5]

PoweredUNKNOWN          42
Powered755BVAK3ZNJYW    23
PoweredXZH7USFBU76DX    21
PoweredOKXVAKQFFLXU5    20
PoweredZ5IMPCBKO5P7E    19
                        ..
Powered870692            6
PoweredYDQW5ISWTRM6T     6
PoweredD48BF             6
Powered482423            6
PoweredIDFLS4WD6GVYV     6
Name: sumdID, Length: 235, dtype: int64

193 scooters, or 9%, performed well with at least 6 rides that day.

In [29]:
#make a list of datetimes with datetime function 
start = datetime(2019, 6, 24, 0, 0)
end = datetime(2019, 6, 25, 0, 0)
delta = timedelta(days=1)
timelist = []
timenamelist = []
while start <= end:
    timelist.append(start)
    timenamelist.append(start.strftime("%Y-%m-%d-%H-%M"))
    start += delta
# print(timelist)
# print(timenamelist)

In [30]:
#goal of function: using list of time frames, in this case, days, make a dataframe containing
#info about the location for each scooter, for each day


#make dictionary of dataframes using base dataframe and a user-defined segments_of_time_list
#from cell above
def make_day_df(datetime_list, timenames_list, dataframe, colname):
    resultslist = {} #turn into dictionary
    x = 0
    y = 1
    z = 0
    while x in range(0,len(datetime_list)-1):
        starttime = datetime_list[x]
        endtime = datetime_list[y]
        mask = (dataframe[colname] > starttime) & (dataframe[colname] <= endtime)
        framename = dataframe.loc[mask]
        x += 1
        y += 1
        print(x)
        resultslist[timenames_list[z]] = framename
        z += 1
    return resultslist
        
the_output = make_day_df(timelist, timenamelist, june_trip,"pubTimeStamp")


1


In [31]:
the_output.keys()

dict_keys(['2019-06-24-00-00'])

In [32]:
the_output['2019-06-24-00-00']
# the_output['2019-05-01-00-00']['sumdid'].nunique()

pubTimeStamp companyName tripRecordNum          sumdID  \
166573 2019-06-24 00:00:20.473        Bird       BRD5729    PoweredSULRR   
166574 2019-06-24 00:00:20.473        Bird       BRD5723    Powered3I7X6   
166575 2019-06-24 00:00:20.473        Bird       BRD5730    PoweredQDRMQ   
166576 2019-06-24 00:00:20.473        Bird       BRD5727    PoweredKURT7   
166577 2019-06-24 00:00:20.473        Bird       BRD5722    PoweredS4QFL   
...                        ...         ...           ...             ...   
170072 2019-06-24 23:56:23.937        Bird       BRD2576    Powered4BFZX   
170073 2019-06-24 23:56:29.410        Lyft        LFT401   Powered099043   
170074 2019-06-24 23:56:38.647        Lyft        LFT402   Powered047053   
170075 2019-06-24 23:57:51.697        SPIN        SPI228  Powered3005365   
170076 2019-06-24 23:57:51.697        SPIN        SPI223  Powered4664287   

        tripDuration  tripDistance            startDate        startTime  \
166573     42.000000    8530.18400  2019-06-23 00:00:00  23:18:28.056666   
166574      7.000000       0.00000  2019-06-23 00:00:00  23:55:19.193333   
166575     50.000000   19028.87200  2019-06-23 00:00:00  23:11:12.990000   
166576     10.000000       0.00000  2019-06-23 00:00:00  23:51:01.563333   
166577      4.000000       0.00000  2019-06-23 00:00:00  23:56:39.883333   
...              ...           ...                  ...              ...   
170072      2.000000     656.16800  2019-06-24 00:00:00  23:55:06.866666   
170073     17.969400   10980.97148  2019-06-24 00:00:00  23:38:31.123333   
170074     19.119667    8392.38872  2019-06-24 00:00:00  23:37:31.346666   
170075     17.000000    2224.40952  2019-06-24 00:00:00         23:37:35   
170076     30.000000    5636.48312  2019-06-24 00:00:00         23:23:46   

                    endDate          endTime  startLatitude  startLongitude  \
166573  2019-06-24 00:00:00  00:00:21.876666      36.148200      -86.812900   
166574  2019-06-24 00:00:00  00:02:05.743333      36.152200      -86.784600   
166575  2019-06-24 00:00:00  00:00:44.086666      36.148300      -86.813800   
166576  2019-06-24 00:00:00  00:00:41.550000      36.162200      -86.774000   
166577  2019-06-24 00:00:00  00:00:56.276666      36.159100      -86.777300   
...                     ...              ...            ...             ...   
170072  2019-06-24 00:00:00  23:57:03.553333      36.147100      -86.808500   
170073  2019-06-24 00:00:00  23:56:29.286666      36.161650      -86.790160   
170074  2019-06-24 00:00:00  23:56:38.526666      36.161580      -86.789970   
170075  2019-06-24 00:00:00         23:53:53      36.152840      -86.782573   
170076  2019-06-24 00:00:00         23:53:21      36.140731      -86.807791   

        endLatitude  endLongitude  \
166573    36.147800    -86.812800   
166574    36.159500    -86.779700   
166575    36.148500    -86.813600   
166576    36.162400    -86.774600   
166577    36.161000    -86.781300   
...             ...           ...   
170072    36.147100    -86.808500   
170073    36.155640    -86.786570   
170074    36.155680    -86.786590   
170075    36.150299    -86.775717   
170076    36.125294    -86.807451   

                                                tripRoute  \
166573  [[36.148236,-86.812822],[36.148295,-86.812806]...   
166574  [[36.152346,-86.784422],[36.152307,-86.784321]...   
166575  [[36.148366,-86.81371],[36.148344,-86.813683],...   
166576  [[36.162272,-86.773973],[36.16231,-86.773939],...   
166577  [[36.159112,-86.777302],[36.159077,-86.77724],...   
...                                                   ...   
170072                           [[36.147171,-86.808436]]   
170073  [[36.16165,-86.79016],[36.16166,-86.79022],[36...   
170074  [[36.16158,-86.78997],[36.1617,-86.79004],[36....   
170075  [[36.150298852647225,-86.775717128258677],[36....   
170076  [[36.12503,-86.80804],[36.1252943,-86.8074509]...   

                         create_dt  
166573  2019-06-26 16:09:47.

In [33]:
the_output['2019-06-24-00-00']['sumdID'].value_counts()
# the_output['2019-05-01-00-00']['sumdid'].nunique()

PoweredUNKNOWN          20
PoweredLFS2H5VQHIS7Y    11
PoweredLXV7RJSLHE56P     8
PoweredRWCZJHPI2C4ZP     8
PoweredJDCZNIOP6YWNY     7
                        ..
PoweredBF16Y             1
PoweredMCYTI             1
PoweredYFCTB             1
PoweredXTFUW             1
Powered099043            1
Name: sumdID, Length: 1849, dtype: int64

1829 scooters total

In [34]:
the_output['2019-06-24-00-00']['sumdID'].value_counts().loc[lambda x : x<3]

Powered972550                                   2
Powered-2c1292e4-f3b5-fbfd-ca6e-e0c076df5d22    2
Powered4816640                                  2
Powered120639                                   2
Powered868360                                   2
                                               ..
PoweredBF16Y                                    1
PoweredMCYTI                                    1
PoweredYFCTB                                    1
PoweredXTFUW                                    1
Powered099043                                   1
Name: sumdID, Length: 1438, dtype: int64

1438 scooters had less than 3 rides. So only 391 had 3 or more. That's 21%

In [35]:
the_output['2019-06-24-00-00']['sumdID'].value_counts().loc[lambda x : x>5]

PoweredUNKNOWN                                  20
PoweredLFS2H5VQHIS7Y                            11
PoweredLXV7RJSLHE56P                             8
PoweredRWCZJHPI2C4ZP                             8
PoweredJDCZNIOP6YWNY                             7
Powered462547                                    7
PoweredZQ7D6YSOBD3N4                             7
PoweredXX36SEXFTINN6                             7
PoweredVPFB2                                     7
PoweredPKBZ6QO7CIUII                             7
PoweredP7EA7PDZH6H4H                             7
PoweredMT7GMY3GHJMSI                             7
PoweredQVY4J5WUDWHU6                             7
PoweredN6G3CGCX7EQSH                             7
PoweredB5U6I3K64AQLG                             7
PoweredJXWC5DTFBJQRD                             7
PoweredGF4UW4AODW5AF                             7
PoweredNU3TYLT5YIIXW                             7
PoweredRRC5FV6DYYHBL                             6
Powered4EJEZQ3KHHGUT           

45 scooters had 6 or more rides, or 2%